### PR TITLE
sql: support alter table owner to command

### LINKED ISSUE #52020: sql: support the `ALTER TABLE OWNER TO` command
https://www.postgresql.org/docs/9.1/sql-altertable.html

### PR DESCRIPTION
**Pull request by RichardJCai**
Release note (sql change): Support ALTER TABLE <type> OWNER TO <owner>
The command changes the owner of a table.

To use the conditions, the following conditions must be met:
The user executing the command must be the explicit owner of the table.
(Not just have membership of the role who owns the table.)
The user executing the command must be a member of the new owner role.
The new owner role must have CREATE on the schema the table belongs to.

Fixes https://github.com/cockroachdb/cockroach/issues/52020

### CODE REVIEW

**Review by rohany — CHANGES_REQUESTED** [review]
Going to just cc @knz for a peek on permissions related PRs.

File: pkg/sql/alter_table.go
Comment (RichardJCai) [reply pkg/sql/alter_table.go:1194]: Should I add that now or wait until your PR is merged?
I wouldn't mind waiting for that PR since a lot of that work will be helpful for ALTER SCHEMA OWNER as well, can do the refactor there.

Comment (RichardJCai) [reply pkg/sql/alter_table.go]: Updated.

**Review by solongordon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52659)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @RichardJCai, and @rohany)

___
*[pkg/sql/alter_table.go, line 742 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhKOE43iRkFnZPGLtu:-MEhKOE43iRkFnZPGLtv:brcvqgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L742)):*
> ```Go
> 				return err
> 			}
> 			descriptorChanged = changed
> ```

I think you want `descriptorChanged = descriptorChanged || changed` here. We don't want to flip it from true to false.

Actually it doesn't look like you're the first one to make this mistake 🙈. I'll open a separate PR to fix some of the other cases.
___
*[pkg/sql/alter_table.go, line 1170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhR_z05-09JNwjlH3k:-MEhR_z05-09JNwjlH3l:boz7htw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1170)):*
> ```Go
> 	}
> 	if !roleExists {
> 		return false, pgerror.Newf(pgcode.UndefinedObject, "role/user %s does not exist", newOwner)
> ```

Nit: Can you put the role name in quotes like postgres does?
___
*[pkg/sql/alter_table.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhOKwBD7orCqsp85p5:-MEhOKwBD7orCqsp85p6:bc4zvza) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1174)):*
> ```Go
> 
> 	// The user explicitly has to be owner to change the owner.
> 	if ok := IsOwner(desc, p.User()); !ok {
> ```

I don't think this is the case. It seems like you just need to have ownership, which you already check for at the beginning of ALTER TABLE. I tried it out in the Postgres and it seems fine to just be a member of the owning role.

It would be weird if you had to be the explicit owner because then if the table were owned by a role, only admins would be able to change the owner.

(Which makes me realize that this check also prevents admins from changing the owner!)
___
*[pkg/sql/alter_table.go, line 1189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhRV4w0raSxGUhyuj8:-MEhRV4w0raSxGUhyuj9:boz7htw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1189)):*
> ```Go
> 	if _, ok := memberOf[newOwner]; !ok {
> 		return false, pgerror.Newf(
> 			pgcode.InsufficientPrivilege, "must be member of role %s", newOwner)
> ```

Nit: Can you put the role name in quotes like postgres does?
___
*[pkg/sql/alter_table.go, line 1194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhP11nDWYjFOuA_jQ5:-MEhOwvL01Nissr72ByM:b-srvxgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1194)):*
> ```Go
> 	// Ensure the new owner has CREATE privilege on the table's schema.
> 	// We do not have to check for the public schema.
> 	if err := p.canCreateOnSchema(ctx, desc.GetParentSchemaID(), newOwner); err != nil {
> ```

Does this handle admins who don't have explicit CREATE privileges? Let's make sure to have a test for that.
___
*[pkg/sql/authorization.go, line 478 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhSp09B2g4QjVuCJEh:-MEhSp09B2g4QjVuCJEi:b-nrnr2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/authorization.go#L478)):*
> ```Go
> 		return nil
> 	case sqlbase.SchemaTemporary, sqlbase.SchemaVirtual:
> 		return errors.Newf("cannot CREATE on schema with id %d", resolvedSchema.ID)
> ```

Can we use a name rather than an ID here?
___
*[pkg/sql/catalog/resolver/resolver.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhR96Y0yGINVnmpUy6:-MEhR96ZAB9tgKQXL1LP:b-w5p82v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/catalog/resolver/resolver.go#L278)):*
> ```Go
> 
> // ResolveSchemaDescByID resolves a schema based on it's ID.
> func ResolveSchemaDescByID(
> ```

@rohany Can you confirm this function looks good and isn't duplicating logic from elsewhere?
___
*[pkg/sql/catalog/resolver/resolver.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhQnFy3J14wqOmDF94:-MEhQnFy3J14wqOmDF95:bnqhnah) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/catalog/resolver/resolver.go#L288)):*
> ```Go
> 	}
> 
> 	// We have already considered public and virtual schemas don't have ids.
> ```

Where did you consider virtual schemas?
___
*[pkg/sql/logictest/testdata/logic_test/alter_table_owner, line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhRpnA1GWsO_36gJwS:-MEhRpnA1GWsO_36gJwT:b8qka44) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/logictest/testdata/logic_test/alter_table_owner#L8)):*
> ```
> ALTER TABLE t OWNER TO fake_user
> 
> # Ensure user the current user is a member of the role we're setting to.
> ```

Nit: "user the current user"
___
*[pkg/sql/logictest/testdata/logic_test/alter_table_owner, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhRsJ-3QWxNFRyjRr2:-MEhRsJ-3QWxNFRyjRr3:bfw761s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/logictest/testdata/logic_test/alter_table_owner#L14)):*
> ```
> user testuser
> 
> # Ensure the user has to be an owner alter the owner.
> ```

Nit: "an owner alter the owner"

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52659)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_table.go, line 1181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEVDDez3wBU4d0HC2lm-r1-1181:-MEii36_34iw4GZHOQvf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a63e6c9d6a91aa97e37590059b30236a0745c5/pkg/sql/alter_table.go#L1181)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: comment longer than 80 chars
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 1194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEVDDez3wBU4d0HC2lm-r1-1194:-MEiiakA7cKEPEifQp_n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a63e6c9d6a91aa97e37590059b30236a0745c5/pkg/sql/alter_table.go#L1194)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Updated.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 742 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhKOE43iRkFnZPGLtu:-MEihbarClulLs0vkDsr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L742)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think you want `descriptorChanged = descriptorChanged || changed` here. We don't want to flip it from true to false.

Actually it doesn't look like you're the first one to make this mistake 🙈. I'll open a separate PR to fix some of the other cases.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 1174 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhOKwBD7orCqsp85p5:-MEifXkM5kGyYQYPbS-u:bnke1ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1174)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I don't think this is the case. It seems like you just need to have ownership, which you already check for at the beginning of ALTER TABLE. I tried it out in the Postgres and it seems fine to just be a member of the owning role.

It would be weird if you had to be the explicit owner because then if the table were owned by a role, only admins would be able to change the owner.

(Which makes me realize that this check also prevents admins from changing the owner!)
</blockquote></details>

Ah I just double checked and this is the case, I thought I had a case where it being just part of an owner role didn't work.
___
*[pkg/sql/alter_table.go, line 1194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhP11nDWYjFOuA_jQ5:-MEj-iCT5aN9tsfE_GXu:b9zi9lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/alter_table.go#L1194)):*
> I added the logic but this check should never happen for virtual schemas, I'm wondering if it should be a panic instead.
It should since it uses the regular CheckPrivilege function but I added a test as well
___
*[pkg/sql/authorization.go, line 478 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhSp09B2g4QjVuCJEh:-MEijLY3-SocKoXdmgMk:b-k835gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/authorization.go#L478)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Can we use a name rather than an ID here?
</blockquote></details>

I'm actually updating this with an assertion error - a virtual schema should not be found here since they don't have ids. 
@rohany just wanna double check that what I said makes sense.
___
*[pkg/sql/catalog/resolver/resolver.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhQnFy3J14wqOmDF94:-MEimFij8P5ifDWt8Ssx:bwi490o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/catalog/resolver/resolver.go#L288)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Where did you consider virtual schemas?
</blockquote></details>

I'm actually a bit confused about this case. I thought virtual schemas can't have ids, but looking at resolver.go staticSchemaIDMap there is a mapping from some schema ids to "virtual schemas" 

I added the logic but this check should never happen for virtual schemas, I'm wondering if it should be a panic/err instead.

**Review by solongordon — APPROVED** [review]
:lgtm: pending another look from @rohany.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-:-MErZ_ra2PTmTwYMpdQG:b-21sdgp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @RichardJCai, and @rohany)

___
*[pkg/sql/catalog/resolver/resolver.go, line 288 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52659#-MEhQnFy3J14wqOmDF94:-MErZJOW1X-y2NaCYcGI:b-etlozm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecba9b9770e8ee18d149ff43f5d7db6f56c78f0d/pkg/sql/catalog/resolver/resolver.go#L288)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I'm actually a bit confused about this case. I thought virtual schemas can't have ids, but looking at resolver.go staticSchemaIDMap there is a mapping from some schema ids to "virtual schemas" 

I added the logic but this check should never happen for virtual schemas, I'm wondering if it should be a panic/err instead.
</blockquote></details>

I'm not sure about this either, @rohany?

File: pkg/sql/alter_table.go
Comment (rohany) [inline pkg/sql/alter_table.go:1191]: this comment isn't needed since the method is handling the public schema check

File: pkg/sql/authorization.go
Comment (rohany) [inline pkg/sql/authorization.go:485]: should this be "within" rather than on?

File: pkg/sql/authorization.go
Comment (rohany) [inline pkg/sql/authorization.go:479]: I think that the virtual schema check should be its own case always returning an error

File: pkg/sql/catalog/resolver/resolver.go
Comment (rohany) [inline pkg/sql/catalog/resolver/resolver.go:278]: I wouldn't call this `ResolveSchemaDescByID` because we aren't gauranteed to return a descriptor always. Just `ResolveSchemaByID` will do. Add some TODO's here for lucy and I to move this into the `descs.Collection`

File: pkg/sql/catalog/resolver/resolver.go
Comment (rohany) [inline pkg/sql/catalog/resolver/resolver.go:308]: I'd rather just not handle the temporary schema case here than assume nil == temp. Leaving the TODO is fine however, but assign it to SQLSchema

**Review by rohany — CHANGES_REQUESTED** [review]

File: pkg/sql/authorization.go
Comment (RichardJCai) [reply pkg/sql/authorization.go:479]: Would it be fine to have the same error message logic for SchemaTemporary and SchemaVirtual? 
Since these cases right now will both just lookup the schema name and return an error for it.

And looking up the schema name for a virtual schema should be okay.

Comment (rohany) [reply pkg/sql/authorization.go:479]: yeah you're right, I think i read this wrong initially. However, I am making a change in https://github.com/cockroachdb/cockroach/pull/52726 that you might want to just copy. I adjusted the `ResolvedSchema` struct to always contain the name of the resolved schema, so that you don't have to look it up again here.

File: pkg/sql/authorization.go
Comment (RichardJCai) [reply pkg/sql/authorization.go:485]: I think I'd actually prefer leaving this as on since we do `GRANT x ON SCHEMA` and our regular check privilege error will say ON for schema as well.

File: pkg/sql/authorization.go
Comment (RichardJCai) [reply pkg/sql/authorization.go:479]: Done

File: pkg/sql/catalog/resolver/resolver.go
Comment (rohany) [inline pkg/sql/catalog/resolver/resolver.go:285]: missing name here

File: pkg/sql/catalog/resolver/resolver.go
Comment (rohany) [inline pkg/sql/catalog/resolver/resolver.go:294]: and here

File: pkg/sql/catalog/resolver/resolver.go
Comment (rohany) [inline pkg/sql/catalog/resolver/resolver.go:326]: you can just do desc.Name, why look up the name again with resolveschemanamebyID?

**Review by rohany — CHANGES_REQUESTED** [review]

File: pkg/sql/catalog/resolver/resolver.go
Comment (RichardJCai) [reply pkg/sql/catalog/resolver/resolver.go:285]: Done

File: pkg/sql/catalog/resolver/resolver.go
Comment (RichardJCai) [reply pkg/sql/catalog/resolver/resolver.go:295]: Done

File: pkg/sql/catalog/resolver/resolver.go
Comment (RichardJCai) [reply pkg/sql/catalog/resolver/resolver.go]: Done

**Review by rohany — APPROVED** [review]

**Comment by RichardJCai** [comment]
Thanks for the reviews!

**Pull request merged**
